File: neetcode/blind75/two_pointers/max_water_container.py
# https://neetcode.io/problems/max-water-container

class Solution:
    def maxArea(self, heights: List[int]) -> int:
        if len(heights) < 2:
            return 0

        curr_max = 0
        left = 0
        right = len(heights) - 1

        while left < right:
            curr_max = max(curr_max, (right - left) * min(heights[left], heights[right]))

            # keep the tallest and most the shorter height pointer
            if (heights[left] < heights[right]):
                left += 1
            else:
                right -= 1

        return curr_max

File: neetcode/neet250_JS/Two-Pointers/max-water-container.js
// https://neetcode.io/problems/max-water-container

/*
max amount of water is the area that is created by the (min height of the two sides) * the distance

left pointer = 0
right pointer = heights.length - 1

track max water

while (left < right) {
    max = Math.max(max, Math.min(heights[left], heights[right]) * (right - left)))

    // keep the highest height
    if (height[left] > height[right]) {
        right -= 1
    } else {
        left += 1
    }
}

return max

- Time: O(n)
- Space: O(1)
*/

class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    maxArea(heights) {
        if (heights.length < 2) {
            return 0
        }

        let max = 0

        let left = 0
        let right = heights.length - 1
        while (left < right) {
            max = Math.max(max, Math.min(heights[left], heights[right]) * (right - left))

            if (heights[left] > heights[right]) {
                right -= 1
            } else {
                left += 1
            }
        }

        return max
    }
}
